File: Level.2/조이스틱.cpp
#include <string>
#include <vector>

using namespace std;

int change(char c) {
    return min(c - 'A', 'Z'- c + 1);
}

int solution(string name) {
    int answer = 0, pos = 0;
    int len = name.length();
    string As(len, 'A');
    for(int i = 0; i < len; ++i) {
        answer += change(name[i]);
    }
    
    while(name != As) {
        name[pos] = 'A';
        int right = pos, left = pos;
        int r_cnt = 0, l_cnt = 0;
        if(name == As) break;
        while(name[right] == 'A' && r_cnt <= len) {
            if(right == len - 1) right = 0;
            else ++right;
            ++r_cnt;
        }
        while(name[left] == 'A' && l_cnt <= len) {
            if(left == 0) left = len - 1;
            else --left;
            ++l_cnt;
        }
        
        if(l_cnt < r_cnt){
            
            pos = left;
            answer += l_cnt;
        }
        else{
            pos = right;
            answer += r_cnt;
        }
    }
    
    return answer;
}

File: JavaScript/Level.2/조이스틱.js
function solution(name) {
    var answer = 0;
    let n = name.length;
    let diffIndex = [];
    for (let i = 0; i < n; ++i) {
        if (name[i] !== "A") {
            diffIndex.push(i);
            answer += Math.min(
                name[i].charCodeAt() - 65,
                90 - name[i].charCodeAt() + 1
            );
        }
    }
    let diffLen = diffIndex.length;
    let minLen = 30;
    let visited = Array.from({ length: diffLen });
    function DFS(index, depth, len) {
        if (depth === diffLen) {
            if (len < minLen) {
                minLen = len;
            }
            return;
        }
        for (let i = 0; i < diffLen; ++i) {
            if (!visited[i]) {
                visited[i] = 1;
                let minValue = Math.min(
                    Math.abs(index - diffIndex[i]),
                    n - Math.abs(index - diffIndex[i])
                );
                DFS(diffIndex[i], depth + 1, len + minValue);
                visited[i] = 0;
            }
        }
    }
    DFS(0, 0, 0);
    answer += minLen;
    return answer;
}
